identity-android/src/main/java/com/android/identity/SessionEncryption.java

```diff
@@ -51,50 +53,72 @@
 
-    private static final String TAG = "SessionEncryptionReader";
+    private static final String TAG = "SessionEncryption";
+
+    public static final int ROLE_MDOC = 0;
+    public static final int ROLE_MDOC_READER = 1;
+
+    @Retention(RetentionPolicy.SOURCE)
+    @IntDef(value = {ROLE_MDOC, ROLE_MDOC_READER})
+    public @interface Role {
+    }
 
+    protected final @DataTransport.Role int mRole;
     private boolean mSessionEstablishmentSent;
 
-    private final PrivateKey mEReaderKeyPrivate;
+    private final PrivateKey mESenderKeyPrivate;
     private final PublicKey mEReaderKeyPublic;
-    private final PublicKey mEDeviceKeyPublic;
+    private final PublicKey mEReceiverKeyPublic;
 
-    private final SecretKeySpec mSKDevice;
-    private final SecretKeySpec mSKReader;
-    private int mSKDeviceCounter = 1;
-    private int mSKReaderCounter = 1;
+    private final SecretKeySpec mSKReceiver;
+    private final SecretKeySpec mSKSender;
+    private int mSKReceiverCounter = 1;
+    private int mSKSenderCounter = 1;
     private boolean mSendSessionEstablishment = true;
 
     /**
-     * Creates a new {@link SessionEncryptionReader} object.
+     * Creates a new {@link SessionEncryption} object. eReaderKeyPublic is required if you are
+     * making a reader {@link SessionEncryption} object (ie. role is ROLE_MDOC_READER).
      *
-     * @param eReaderKeyPrivate the reader private ephemeral key.
-     * @param eReaderKeyPublic the reader public ephemeral key.
-     * @param eDeviceKeyPublic the device public key.
-     * @param encodedSessionTranscript the bytes of the <code>SessionTranscript</code> CBOR.
+     * <p>The <code>DeviceEngagement</code> and <code>Handover</code> CBOR referenced in the
+     * parameters below must conform to the CDDL in ISO 18013-5.
+     *
+     * @param role The role of the SessionEncryption object
+     * @param eSenderKeyPrivate The private ephemeral key of the sender (e.g. if ROLE_MDOC_READER,
+     *                          then the sender is the reader and the receiver is the holder).
+     * @param eReceiverKeyPublic The public ephemeral key of the receiver.
+     * @param encodedSessionTranscript The bytes of the <code>SessionTranscript</code> CBOR.
+     * @param eReaderKeyPublic The reader public ephemeral key. Not required if the role is ROLE_MDOC.
      */
-    public SessionEncryptionReader(@NonNull PrivateKey eReaderKeyPrivate,
-            @NonNull PublicKey eReaderKeyPublic,
-            @NonNull PublicKey eDeviceKeyPublic,
-            @NonNull byte[] encodedSessionTranscript) {
-        mEReaderKeyPrivate = eReaderKeyPrivate;
+    public SessionEncryption(@Role int role,
+                             @NonNull PrivateKey eSenderKeyPrivate,
+                             @NonNull PublicKey eReceiverKeyPublic,
+                             @NonNull byte[] encodedSessionTranscript,
+                             @Nullable PublicKey eReaderKeyPublic) {
+        mRole = role;
```

Comment:
The terminology here seems wrong and I think it's confusing to introduce sender/receiver. Conceptually the parameters needed to establish _mdoc session encryption_ with another party is

 - our role (`MDOC` or `MDOC_READER`)
 - a key-pair used on our end
 - a public key from the other end

so I think it should just be

```
public SessionEncryption(@Role int role,
    @NonNull KeyPair keyPair,
    @NonNull PublicKey remotePublicKey,
    @NonNull byte[] encodedSessionTranscript)
```

Note that `SessionEncryptionDevice` just took the private key (all that was needed for ECDH) and `SessionEncryptionReader` took a `PublicKey` and `PrivateKey` but that should really have been a `KeyPair` instead.